clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -220,33 +220,38 @@ public synchronized TransactionalRequestResult sendOffsetsToTransaction(Map<Topi
     }
 
     public synchronized void maybeAddPartitionToTransaction(TopicPartition topicPartition) {
-        if (!isInTransaction())
-            throw new IllegalArgumentException("Cannot add partitions to a transaction in state " + currentState);
+        if (currentState != State.IN_TRANSACTION)
+            throw new IllegalStateException("Cannot add partitions to a transaction in state " + currentState);
 
-        if (partitionsInTransaction.contains(topicPartition))
+        if (partitionsInTransaction.contains(topicPartition) || pendingPartitionsInTransaction.contains(topicPartition))
             return;
 
         newPartitionsInTransaction.add(topicPartition);
     }
 
-    public RuntimeException lastError() {
+    RuntimeException lastError() {
         return lastError;
     }
 
-    public synchronized boolean ensurePartitionAdded(TopicPartition tp) {
+    public synchronized void failIfUnreadyForSend() {
```

Comment:
Funny. I used that name initially and changed it because I thought the current one read nicer. Now I'm reading it again and I'm inclined to change it back.